### ISSUE TITLE
Can't export to AVIF format

### ISSUE THREAD

**Issue by criadoperez** [darktable-org/darktable#7177]
**Describe the bug**
darktable produces an exiv2 exception when trying to export any image to avif format. 

Things I tried:
 - Export to other formats work fine. Tried jpeg and jpeg2000..
 - Modifying avif setting (lossless/lossy, 8-10-12bit, tiling or quality produces the same result (doesn't work)
 - Tried exporting from jpeg, tiffs and avifs. (all to avifs) Results is always the same.
 - Executing darktable with sudo privileges produces same result.
 - In "edit metada exportation", deselecting all options, so no metadata is exported also produces same result.

**To Reproduce**
1. Open darktable
2. Import an image
3. Export the image to avif format.
4. Error will show in darktable and in the command line, and no file is created.
-->

**Expected behavior**
An AVIF file should be created.

**Screenshots**
alejandro@openair:/opt/darktable/bin$ ./darktable
[defaults] found a 64-bit system with 8088188 kb ram and 4 cores (0 atom based)
[defaults] setting high quality defaults
[dt_exif_xmp_attach_export] /home/alejandro/Desktop/darktable_exported/_MG_6689.avif: caught exiv2 exception '/home/alejandro/Desktop/darktable_exported/_MG_6689.avif: Failed to open the data source: No such file or directory (errno = 2)'
[imageio_storage_disk] could not export to file: `/home/alejandro/Desktop/darktable_exported/_MG_6689.avif'!

**Platform**
 - Darktable Version: 3.3.0+1961~g41ad408e2
 - OS: Linux Ubuntu 20.10
 - Hardware: MacBookAir i5 8GB RAM
 - OpenCL not activated
 - Exiv2 package version 0.27.3-3 (current version on the ubuntu repository)
 - libavif v.0.8.3-3
 - libgav1 v.0.16.1

**Additional context**

**Comment by criadoperez** [comment]
I confirm the same behavior on version 3.3.0+1977~gde6f4f817 with all optional dependencies installed (except GMIC which shouldn't affect at all)
Can anybody else confirm the issue? Thanks in advance.

**Comment by criadoperez** [comment]
Using the command line produces the same error.

alejandro@openair:~$ /opt/darktable/bin/darktable-cli ~/Desktop/_MG_6689.tif cooltest.avif --verbose
orientation (on)
sharpen (on)
display encoding (on)
output color profile (on)
input color profile (on)
[dt_exif_xmp_attach_export] cooltest.avif: caught exiv2 exception 'cooltest.avif: Failed to open the data source: No such file or directory (errno = 2)'
[imageio_storage_disk] could not export to file: `cooltest.avif'!
alejandro@openair:~$

**Comment by criadoperez** [comment]
This seems to be an EXIV2 issue so I will be closing the issue here.
The latest release today of EXIV2 is 0.27.3. According to the [EXIV2 github site](https://github.com/Exiv2/exiv2/milestone/6),  support of ISMOBMFF Files (AVIF, HEIF, CR3) is in the milestone for v. 0.27.4 with current expected release date June 30th, 2021.
Issue details are also [here](https://github.com/Exiv2/issues/1229).

**Comment by parafin** [comment]
I'm lost here, if it doesn't work at all, how and why it was added as a feature to dt?

**Comment by criadoperez** [comment]
Good question parafin. :-)

Also, if its only a metadata issue, darktable should in theory still be able to export to avif if you decide to export without metadata. However if in "edit metadata exportation", you deselect all options, so no metadata is exported, it still fails. Is it still trying to export some metadata even if you deselect all options?

**Comment by kmilos** [comment]
Can you run with debug options please, there might be more clues?

It doesn't look like libavif uses Exiv2 to store metadata, but it's own methods, so perhaps this started happening when libavif version was changed to >=0.8.2 in October or so...

**Comment by criadoperez** [comment]
Good idea @kmilos Here it is.

[darktable-debug-output.txt](https://github.com/darktable-org/darktable/files/5666694/darktable-debug-output.txt)

I run the following command:

`/opt/darktable/bin/darktable -d all | tee darktable-debug-output.txt`
Then selected a tiff file, exported the file to tiff and as soon as the error appeared I closed darktable. I attach the debug output of these steps.
It might be important to mention that in this case, I left the default "metadata export options" as they are by default (which exports with metadata).
In lines 2180 to 2186 of the output file I'm surprised to see this text:

```
79,186782 Exporting AVIF image [/home/alejandro/Desktop/darktable_exported/_MG_6689.avif] [width: 8688, height: 5792, bit depth: 12, comp: lossy, quality: 50]
80,898424 [camera_control] loaded 39 port drivers.
80,925599 [camera_control] 0 cameras connected
80,969288 [avif colorprofile profile: (null) - icc]
80,969369 [sql] /home/alejandro/mycompiled_darktable/darktable-master/src/common/colorspaces.c:912, function dt_colorspaces_get_output_profile(): prepare "SELECT op_params FROM main.history WHERE imgid=?1 AND operation='colorout' ORDER BY num DESC LIMIT 1"
80,969791 [avif quality: 50 => maxQuantizer: 50, minQuantizer: 14, tileColsLog2: 0, tileRowsLog2: 0, threads: 4]
80,969822 Failed to encode AVIF image [/home/alejandro/Desktop/darktable_exported/_MG_6689.avif]: No codec available
```
Could it be that this bug has nothing to do with metadata or EXIV2 after all?
Not sure if this is related but, libavif v.0.8.3-3 binary avifenc, allows encoding avif files ONLY from jpeg/png/yum. darktable also fails when you try to export png or jpeg to avif anyway, but how then can darktable encode "anything else" to avif?

```
alejandro@openair:~$ avifenc 
Syntax: avifenc [options] input.[jpg|jpeg|png|y4m] output.avif
...
Version: 0.8.3 (dav1d [dec]:0.7.1, libgav1 [dec]:0.16.1)
libyuv : unavailable
```




Version of darktable for this test: 3.3.0+1977~gde6f4f817

**Comment by kmilos** [comment]
> Could it be that this bug has nothing to do with metadata or EXIV2 after all?

Yep, it looks like it fails even creating an initial AVIF file, and then there is nothing to attach XMP metadata to (the "No such file or directory (errno = 2)").

**Comment by kmilos** [comment]
@cryptomilk Do we also need to perhaps update the write side of things for libavif >0.8.2 API change?

**Comment by cryptomilk** [comment]
I've just compiled and tested with libavif 0.8.3:

```
./bin/darktable
[dt_exif_xmp_attach_export] /home/asn/Pictures/RAW/2019/2019-06_Kletterfestival/2019-06_RAW_Kletterfestival/../2019-06_JPG_Kletterfestival/2019-06_Kletterfestival_0001.avif: caught exiv2 exception '/home/asn/Pictures/RAW/2019/2019-06_Kletterfestival/2019-06_RAW_Kletterfestival/../2019-06_JPG_Kletterfestival/2019-06_Kletterfestival_0001.avif: The file contains data of an unknown image type'
[export_job] exported to `/home/asn/Pictures/RAW/2019/2019-06_Kletterfestival/2019-06_RAW_Kletterfestival/../2019-06_JPG_Kletterfestival/2019-06_Kletterfestival_0001.avif'
```

Works just fine. Will try 0.8.4 next.

**Comment by cryptomilk** [comment]
Tested with libavif 0.8.4 and works just fine:

```
./bin/darktable -d imageio
6.689476 [dt_imageio_export] imgid 21061, pipe 6000x4000, range 0x0 --> exact 0, upscale 0, corrected 0, scale 1.0000000, corr 1.000000, size 6000x4000
7.035770 Exporting AVIF image [/home/asn/Pictures/RAW/2019/2019-06_Kletterfestival/2019-06_RAW_Kletterfestival/../2019-06_JPG_Kletterfestival/2019-06_Kletterfestival_0001_01.avif] [width: 6000, height: 4000, bit depth: 8, comp: lossy, quality: 80]
7.270984 [avif colorprofile profile: sRGB - nclx]
7.271014 [avif quality: 80 => maxQuantizer: 20, minQuantizer: 0, tileColsLog2: 0, tileRowsLog2: 0, threads: 24]
[dt_exif_xmp_attach_export] /home/asn/Pictures/RAW/2019/2019-06_Kletterfestival/2019-06_RAW_Kletterfestival/../2019-06_JPG_Kletterfestival/2019-06_Kletterfestival_0001_01.avif: caught exiv2 exception '/home/asn/Pictures/RAW/2019/2019-06_Kletterfestival/2019-06_RAW_Kletterfestival/../2019-06_JPG_Kletterfestival/2019-06_Kletterfestival_0001_01.avif: The file contains data of an unknown image type'
[export_job] exported to `/home/asn/Pictures/RAW/2019/2019-06_Kletterfestival/2019-06_RAW_Kletterfestival/../2019-06_JPG_Kletterfestival/2019-06_Kletterfestival_0001_01.avif'
```

**Comment by cryptomilk** [comment]
Note that exiv2 is not able to read ISOMBFF containers yet. We need to fix reading the exif data with exiv2 too, we should just hand what we already have in memory to exiv2.

**Comment by criadoperez** [comment]
@cryptomilk it's great to see it works for someone! Thank you for testing. You are the first person that confirmed it works for them and you seem to be an exception.  Maybe now we can figure out why it doesn't work for the rest.
You tested with different versions of libavif, but what version of darktable and exiv2 are you using? What OS?
If you type avifenc, what version of dav1d, libgav1 and libyuv are you using?
Thanks for your input.

**Comment by cryptomilk** [comment]
Run `ldd /usr/bin/darktable | grep libavif` and paste the output here.

**Comment by kmilos** [comment]
I just built and managed to export as well on mingw64 w/ libavif 0.8.2

```
$ /opt/darktable-test/bin/darktable -d imageio --configdir "C:/msys64/home/milos/.config/darktable-test"
168.475332 [dt_imageio_export] imgid 1, pipe 6016x4016, range 0x0 --> exact 0, upscale 0, corrected 0, scale 1.0000000, corr 1.000000, size 6016x4016
169.712282 Exporting AVIF image [C:\temp/darktable_exported/_DSC0983.avif] [width: 6016, height: 4016, bit depth: 12, comp: lossy, quality: 85]
170.226258 [avif colorprofile profile: (null) - icc]
170.227231 [avif quality: 85 => maxQuantizer: 15, minQuantizer: 0, tileColsLog2: 0, tileRowsLog2: 0, threads: 8]
```

(Slow though on my machine, took ~5min for a 24MP image...)

**Comment by criadoperez** [comment]
```
alejandro@openair:~$ ldd /opt/darktable/bin/darktable | grep libavif
	libavif.so.8 => /lib/x86_64-linux-gnu/libavif.so.8 (0x00007f7bd26bc000)
```

**Comment by cryptomilk** [comment]
@criadoperez This looks like a Debian based system. It works fine openSUSE and Fedora.

`ldd /lib/x86_64-linux-gnu/libavif.so.8`

**Comment by criadoperez** [comment]
@kmilos  Whats your OS?
Maybe its a debian based os issue. The other user that reported a problem was also using Ubuntu.

**Comment by kmilos** [comment]
Windows 10, built w/ mingw64

**Comment by criadoperez** [comment]
Great! Maybe we are getting somewhere. I'll install a clean Fedora OS and test to confirm if this is a Debian based issue.

**Comment by ptilopteri** [comment]
I just successfully exported a 24mb image to avif and it pegged
my 12-core for about 3.5 minutes

/usr/lib64/libavif.so.7
openSUSE Tumbleweed 20201207
NVIDIA GF106 [GeForce GTS 450], 390.138
darktable 3.3.0+2052~g7beaa95ab
bit depth is 64 bit
normal build
SSE2 optimized codepath enabled
OpenMP support enabled
OpenCL support enabled
Lua support enabled, API version 6.1.0
Colord support enabled
gPhoto2 support enabled
GraphicsMagick support enabled
ImageMagick support disabled
OpenEXR support enabled
OpenCL on
i7 12-core 36GB
5.9.12-1-default

**Comment by ptilopteri** [comment]
I just successfully exported a 24mb image to avif and it pegged
my 12-core for about 3.5 minutes

/usr/lib64/libavif.so.7
openSUSE Tumbleweed 20201207
NVIDIA GF106 [GeForce GTS 450], 390.138
darktable 3.3.0+2052~g7beaa95ab
bit depth is 64 bit
normal build
SSE2 optimized codepath enabled
OpenMP support enabled
OpenCL support enabled
Lua support enabled, API version 6.1.0
Colord support enabled
gPhoto2 support enabled
GraphicsMagick support enabled
ImageMagick support disabled
OpenEXR support enabled
OpenCL on
i7 12-core 36GB
5.9.12-1-default

and following control line msg:
[dt_exif_xmp_attach_export] /paka/xhome/Pictures/2020/candids/2020.1113a.play/DrkTbl/201113_153614_dse_1387.avif: caught exiv2 exception '/paka/xhome/Pictures/2020/candids/2020.1113a.play/DrkTbl/201113_153614_dse_1387.avif: The file contains data of an unknown image type'

**Comment by kmilos** [comment]
It looks like there aren't any AV1 encoders (libaom or rav1e) pulled in by libavif on Debian-based systems? A packaging oversight, they seem to be required by libavif on Fedora and mingw64...?

@criadoperez Can you try installing either of these manually?

**Comment by parafin** [comment]
I don't know the design of libavif, but most likely used libraries are determined at compile time, so installing anything won't affect it's behaviour outside of "loads" or "not loads". So it would be indeed a packaging oversight, but which requires re-compilation of the package.

**Comment by kmilos** [comment]
From the [Ubuntu package changelog](https://launchpadlibrarian.net/510298790/libavif_0.8.4-1_0.8.4-2.diff.gz):

> + Due to the temporary lack of support libraries, no decoder is available
> + for now. As a result, the avifenc binary won't work.

Should say "encoder", but you get the gist, it seems to be a known problem...

**Comment by parafin** [comment]
I would say then avif export code in darktable should handle such error case more gracefully. Dunno, show a error on export? Don't allow exporting at all?

**Comment by criadoperez** [comment]
In Ubuntu, you can encode avif files with the avifenc tool on the command line. Ubuntu 20.10 only has libavif 0.7 (lower than the 0.8.2 minimum requirement of darktable). However Ubuntu hirsute (future version of Ubuntu) has [libavif 0.8.4](https://packages.ubuntu.com/search?suite=hirsute&searchon=names&keywords=libavif), so I changed temporarily the repositories to hirsute to install the package and it works fine. 
In Ubuntu (with libavif 0.8.3)  the avifenc states: Version 0.8.3 (dav1d [dec]:0.7.1, libgav1 [dec] 0.16.1)
If Fedora the avifenc stats: Version 0.8.2 (dav1d:0.7.1, aom:v2.0.0, rav1e:0.3.3)
So yes, no aom in Ubuntu. Is no aom  the problem for debian systems?

I also installed a clean FEDORA 33 sandbox with all dependencies from [wiki](https://github.com/darktable-org/darktable/wiki/Building-darktable#fedora), compiling from master branch and yes darktable can export to AVIF here but I don't think it does it correctly all the time.
If you try to export from png to avif it works fine. But if you try to export a tiff to avif, it does export a correct avif file, but takes a lot less (seconds instead of minutes) and produces a very low quality image (70 KBs from a 50MP image at 95 of quality).
Its important to note that avifenc currently only supports creating an avif file from  jpeg/png/y4m. How can then darktable export "anything" to avif? 

@kmilos and  @ptilopteri can you please test exporting from a tiff to avif to see if you have the same problem? When you reported success exporting to avif, what file did you export from?

**Comment by kmilos** [comment]
I exported a raw NEF file, a 24MP image ended up being ~8MB at 85% quality.

dt doesn't use avifenc - it is just a frontend CLI tool of libavif to the various AV1 encoders, that happens to only parse a limited number of input formats...

**Comment by kmilos** [comment]
> In Ubuntu (with libavif 0.8.3) the avifenc states: Version 0.8.3 (dav1d [dec]:0.7.1, libgav1 [dec] 0.16.1)
> If Fedora the avifenc stats: Version 0.8.2 (dav1d:0.7.1, aom:v2.0.0, rav1e:0.3.3)
> So yes, no aom in Ubuntu. Is no aom the problem for debian systems?

Yes. As mentioned above, Debian/Ubuntu don't ship any AV1 encoders (aom or rav1e) that are necessary to create an AVIF, only decoders (dav1d, libgav1) to read them.

**Comment by ptilopteri** [comment]
note: renamed to xxx.avif.txt

[p7280827.avif.txt](https://github.com/darktable-org/darktable/files/5673299/p7280827.avif.txt)

**Comment by johnny-bit** [comment]
Should be fixed by #7260 (wit addition of #7375).

If it's still a problem on master (the fix hasn't been released yet as a stable package) please reopen :)
